nwapp/utils/notifier.js

```diff
@@ -28,14 +40,112 @@ function playNotificationSound() {
   audio = null;
 }
 
-module.exports = function (spec) {
-  if (!settings.showNotifications) return;
-  playNotificationSound();
 
-  var notification = new window.Notification(spec.title, {
-    body: spec.message,
-    icon: getNotificationImageForUri(spec.link)
+// Cache lookup
+var urlFilePathCacheMap = new Map();
+var tempAvatarDirectory = path.join(os.tmpDir(), 'gitter-notification-avatars_');
+log.info('Saving notification avatars at: ' + tempAvatarDirectory);
+var ensureTempAvatarDirectory = ensureDir(tempAvatarDirectory)
+  .then(function() {
+    return tempAvatarDirectory;
   });
+events.on('app:quit', function() {
+  log.info('App is quiting, cleanup avatars');
+  rimraf(tempAvatarDirectory)
+    .then(function() {
+      log.info('Finished avatar cleanup');
+    });
+});
+function getFileAtUrl(url) {
+  return ensureTempAvatarDirectory.then(function() {
+      var cachedFilePath = urlFilePathCacheMap.get(url);
+      if(!cachedFilePath) {
+        throw new Error('Avatar is not in the cache');
+      }
+      return cachedFilePath;
+    })
+    .catch(function(cachedFilePath) {
+      return new Promise(function(resolve, reject) {
+        var tempFileStream = temp.createWriteStream({
+          dir: tempAvatarDirectory
+        });
+
+        tempFileStream
+          .on('finish', function() {
+            var tempFilePath = tempFileStream.path;
+            // Store it on our map as a cache lookup
+            urlFilePathCacheMap.set(url, tempFilePath);
+            resolve(tempFilePath);
+          })
+          .on('error', function(err) {
+            reject(err);
+          });
+
+        request({
+            url: url,
+            method: 'GET',
+            timeout: 750
+          })
+          .on('error', function(err) {
+            reject(err);
+          })
+          .pipe(tempFileStream);
+      });
+    })
+    .then(function(filePath) {
+      return 'file://' + filePath;
+    });
+}
+
+
+var notifierDefaults = {
+  title: '',
+  message: '',
+  link: undefined,
+  click: undefined
+};
+
+module.exports = function (options) {
```

Comment:
I disagree with this as they are optional. I could easily see a notification with just a title or message.